File: jcrm-ui/src/app/resources/phone-number-resource.ts
import {PhoneNumber} from '@app/entity/phone-number';

export class PhoneNumberResource {

    public static of(phoneNumber: PhoneNumber): PhoneNumberResource {
        return new PhoneNumberResource(
            phoneNumber.countryCode,
            phoneNumber.regionCode,
            phoneNumber.phoneNumber,
            PhoneNumber.getPhoneTypeId(phoneNumber.type)
        );
    }

    public static toPhoneNumber(resource: PhoneNumberResource): PhoneNumber {
        return new PhoneNumber(
            resource.countryCode,
            resource.regionCode,
            resource.phoneNumber,
            PhoneNumber.getPhoneTypeById(resource.type)
        );
    }

    constructor(
        public readonly countryCode: string,
        public readonly regionCode: string,
        public readonly phoneNumber: string,
        public readonly type: number
    ) {}
}

File: jcrm_ui_flutter/lib/src/jcrm/ui/flutter/resource/phone_number_resource.dart
import 'package:jcrm_ui_flutter/src/jcrm/ui/flutter/entity/phone_number.dart';
import 'package:jcrm_ui_flutter/src/jcrm/ui/flutter/resource/entity_resource.dart';
import 'package:jcrm_ui_flutter/src/jcrm/ui/flutter/resource/json_resource.dart';

class PhoneNumberResource extends EntityResource {

  static List<PhoneNumberResource> fromJsonList(List<dynamic> json) {
    return JsonResource.fromJsonList(json, (toRead) => PhoneNumberResource.fromJson(toRead));
  }

  final String countryCode;
  final String regionCode;
  final String phoneNumber;

  final int type;

  PhoneNumberResource.fromEntity(PhoneNumber phoneNumber)
      : countryCode = phoneNumber.countryCode,
        regionCode = phoneNumber.regionCode,
        phoneNumber = phoneNumber.phoneNumber,
        type = phoneNumber.type;

  PhoneNumberResource.fromJson(Map<String, dynamic> json)
      : countryCode = json['countryCode'],
        regionCode = json['regionCode'],
        phoneNumber = json['phoneNumber'],
        type = json['type'];

  PhoneNumberResource(this.countryCode, this.regionCode, this.phoneNumber, this.type);

  @override
  void buildJson(Map<String, dynamic> json) {
    json['countryCode'] = countryCode;
    json['regionCode'] = regionCode;
    json['phoneNumber'] = phoneNumber;
    json['type'] = type;
  }
}
